components/raftstore/src/store/peer.rs

```diff
@@ -1026,28 +1028,34 @@ where
             // As another role know we're not missing.
             self.leader_missing_time.take();
         }
-        // Here we hold up MsgReadIndex. If current peer has valid lease, then we could handle the
-        // request directly, rather than send a heartbeat to check quorum.
         let msg_type = m.get_msg_type();
-        let committed = self.raft_group.raft.raft_log.committed;
-        let expected_term = self.raft_group.raft.raft_log.term(committed).unwrap_or(0);
-        if msg_type == MessageType::MsgReadIndex && expected_term == self.raft_group.raft.term {
-            // If the leader hasn't committed any entries in its term, it can't response read only
-            // requests. Please also take a look at raft-rs.
-            let state = self.inspect_lease();
-            if let LeaseState::Valid = state {
-                let mut resp = eraftpb::Message::default();
-                resp.set_msg_type(MessageType::MsgReadIndexResp);
-                resp.term = self.term();
-                resp.to = m.from;
-                resp.index = self.get_store().committed_index();
-                resp.set_entries(m.take_entries());
-
-                self.pending_messages.push(resp);
-                return Ok(());
-            }
-            self.should_wake_up = state == LeaseState::Expired;
+        if msg_type == MessageType::MsgReadIndex {
+            if ctx.coprocessor_host.on_step_read_index(&mut m) {
+                return Ok(self.raft_group.step(m)?);
+            }
+            // Here we hold up MsgReadIndex. If current peer has valid lease, then we could handle the
+            // request directly, rather than send a heartbeat to check quorum.
```

Comment:
Even though there are memory locks encountered, I think we can still utilize the leader lease to avoid broadcasting a heartbeat to check quorum. /cc @5kbpers

Reply:
@5kbpers explains that to me. Could you please also take a look at my new commit https://github.com/tikv/tikv/pull/8926/commits/c150ef3ab42449af2b9e11487d1dce4d65e2bfcf?

Reply:
LGTM